Live2D with Lipsync (Seperated)

RaSan147 opened the pull request.

RaSan147 committed:
Added lip-sync module with audio

Changes:
1. updated model.motion(group, index, priority) to model.motion(group, index, priority, sound, volume, expression);
2. added model.stopSpeaking()
3. updated readme.MD with demos
4. Workflow will save files to dist (won't be gitignored)
5. Praying this new change doesn't break

RaSan147 committed:
Automated report

RaSan147 committed:
Update index.js links

RaSan147 committed:
Update test.yml

RaSan147 committed:
Update test.yml

RaSan147 committed:
Update test.yml

RaSan147 committed:
Fix Lip sync. Breaking change

voice volume expressions are now optional arg {name: value, ....}

RaSan147 committed:
Automated report

RaSan147 committed:
Update README.md and added video

guansss reviewed (changes requested):
Sorry for the long wait! You did a great job on this PR and I really appreciate the effort you put into it.

I've added a few comments for things that need to be addressed before this can be merged.

RaSan147 commented on .github/workflows/test.yml:
Well i couldn't run the build in my local machine, so the workflow allowed me build and download. I'll make another branch and and make a pull request.

RaSan147 commented on README.md:
````diff
@@ -97,16 +100,24 @@ import { Live2DModel } from 'pixi-live2d-display/cubism2';
 ```
 
-#### Via CDN
+#### Via CDN (lipsync patched)
````
Yeah will change that.

RaSan147 commented on src/cubism-common/SoundManager.ts:
```diff
@@ -1,7 +1,7 @@
 
 const TAG = 'SoundManager';
-const VOLUME = 0.5;
+const VOLUME = 0.9;
```
However since the lip movement value is dependent on the audio volume, lowering it makes unusual movement (like mumbling)

guansss replied:
Is it possible to add a multiplier after getting the value? Some thing like `amplitude *= 2`

RaSan147 replied:
tried, sending Amp variable to analizer is hard. Let it be...

RaSan147 commented on src/cubism-common/SoundManager.ts:
```diff
@@ -39,6 +41,8 @@ export class SoundManager {
 
         audio.volume = this._volume;
         audio.preload = 'auto';
+        audio.autoplay = true;
```
Thanks for the info

RaSan147 commented on src/cubism2/Cubism2InternalModel.ts:
```diff
@@ -33,9 +34,12 @@ export class Cubism2InternalModel extends InternalModel {
     angleZParamIndex: number;
     bodyAngleXParamIndex: number;
     breathParamIndex: number;
+    mouthFormIndex: number;
```
I do remember it using somewhere in cubism 2, cubism 4 has it by default. It takes the mouthForm id...

RaSan147 commented:
Sorry for the late reply.
I'll make a branch with the changes you mentioned (also fixing a lot of merge conflict from the latest update)
Thanks for taking your time and looking into it. This project is awesome.

RaSan147 commented on src/cubism-common/MotionManager.ts:
```diff
@@ -199,14 +209,125 @@ export abstract class MotionManager<Motion = any, MotionSpec = any> extends Even
+            }
+        }
+        let soundURL: string | undefined;
+        const isBase64Content = sound && sound.startsWith('data:audio/wav;base64');
```
ah thanks. totally missed that.

RaSan147 commented on src/cubism-common/MotionManager.ts:
```diff
@@ -199,14 +209,125 @@ export abstract class MotionManager<Motion = any, MotionSpec = any> extends Even
+        else {
+            soundURL = 'data:audio/wav;base64';
+        }
+        const isUrlPath = sound && (sound.startsWith('http') || sound.startsWith('blob'));
+        let file: string | undefined;
+        if (isUrlPath || isBase64Content) {
+            file = sound;
+        }
+        const that = this;
+        if (file) {
```
Ok, will try to patch up. Please Update the error based on the conversion (this is my 1st work on TS)

RaSan147 commented on src/cubism-common/MotionManager.ts:
```diff
@@ -199,14 +209,125 @@ export abstract class MotionManager<Motion = any, MotionSpec = any> extends Even
+     * @param expression - In case you want to mix up a expression while playing sound (bind with Model.expression())
+     * @returns Promise that resolves with true if the sound is playing, false if it's not
+     */
+    async speakUp(sound: string, {volume=1, expression, resetExpression=true}:{volume?:number, expression?: number | string, resetExpression?:boolean}={}) {
```
I honestly thought they might cause conflict and came up with this name and realized both are different, yet didn't bother to change.

RaSan147 commented on src/cubism-common/MotionManager.ts:
```diff
@@ -199,14 +209,125 @@ export abstract class MotionManager<Motion = any, MotionSpec = any> extends Even
+                this.currentAnalyzer = analyzer;
+
+            } catch (e) {
+                logger.warn(this.tag, 'Failed to create audio', soundURL, e);
```
okk

RaSan147 committed:
Merge branch 'master' into master

RaSan147 committed:
Fix type error on ci test

RaSan147 committed:
clear dist, using workflow to get output files

RaSan147 commented on README.md:
```diff
@@ -121,32 +132,110 @@ import { Live2DModel } from 'pixi-live2d-display';
 // reference window.PIXI.Ticker to automatically update Live2D models
 window.PIXI = PIXI;
 
+var model_proxy; //make a global scale handler to use later
```
![image](https://github.com/guansss/pixi-live2d-display/assets/34002411/afb0fc20-d574-4c76-afd6-ea9fa4b9ff49)
well you used a scoped code to generate model as local variable, So, I assigned it globally to use later 😅

guansss replied:
I mean, generally, examples are supposed to run individually, so it's unnecessary to make example B to reference a variable in example A.
We can simply use `model` in the follow-up examples (just like many examples in the docs site), and users will know how to create `model` because it's already shown in the first example.

Also, I just realized this example should go to the docs (`./docs/docs/`), since the readme is just a place to show very simple and basic examples.
The docs are using in an extended Markdown syntax and requires setting up some Python environment if you want to locally preview them. Later I'll add these instructions to `DEVELOPMENT.md`.

RaSan147 replied:
It'd be really helpful if you add them to the docs (never did docs before), I've written things quite clear and understandable (updating a new branch from your latest comments to fix them, only a few, left, just busy to complete and submit)

guansss replied:
Okay I will do it when you've finished them.

RaSan147 replied:
Yay, thanks 😸

RaSan147 committed:
removed Version number

RaSan147 committed:
Now supports other b64 audio

RaSan147 committed:
Remove audio url validation

RaSan147 committed:
rename speakUp -> speak

RaSan147 committed:
rename resetMotions -> stopMotions

RaSan147 committed:
return false when audio play fails

RaSan147 commented on .gitignore:
deleted

RaSan147 committed:
Added crossorigin for speak voice source

RaSan147 committed:
Place missing CrossOrigin arg

RaSan147 committed:
Update MotionManager.ts

RaSan147 committed:
Update readme as per PR

RaSan147 committed:
Check any base64 data for audio

Co-authored-by: Guan <46285865+guansss@users.noreply.github.com>

RaSan147 committed:
Remove autoplay attr from audio

Co-authored-by: Guan <46285865+guansss@users.noreply.github.com>

RaSan147 committed:
Remove cache blocker.

Co-authored-by: Guan <46285865+guansss@users.noreply.github.com>

RaSan147 committed:
remove wav only blob condition

also remove cache buster and autoplay

RaSan147 commented on src/cubism-common/SoundManager.ts:
```diff
@@ -39,6 +41,8 @@ export class SoundManager {
         audio.volume = this._volume;
         audio.preload = 'auto';
+        audio.autoplay = true;
+        audio.crossOrigin = "anonymous";
```
Fixed

RaSan147 committed:
Merge branch 'guansss:master' into master

RaSan147 committed:
Merge branch 'guansss:master' into for_PR

RaSan147 committed:
RAN `npm run lint:fix`

RaSan147 committed:
Merge branch 'for_PR' of https://github.com/RaSan147/pixi-live2d-display into for_PR

RaSan147 committed:
Merge remote-tracking branch 'origin/for_PR'

RaSan147 committed:
reverting readme with this repo links

RaSan147 committed:
fix startmotion, getting error

RaSan147 committed:
All test passed. Everything OK

guansss committed:
revert: undo unexpected changes

guansss committed:
fix: model.speak() not returning true on success

RaSan147 committed:
Bring patch from for_pr

9a9d20ffde79eecbd99e1eaea21cd0f6aa8ccc32

638053e0385fb9608b78236b07c5aad01cb893d8

RaSan147 committed:
Merge branch 'update'

RaSan147 committed:
Fix artifact upload on failed test

RaSan147 committed:
Update test.yml

RaSan147 committed:
Update test.yml

RaSan147 committed:
Upload dist for jsdelivr support

RaSan147 committed:
Update .gitignore

RaSan147 committed:
Automated report

RaSan147 committed:
Update README.md

RaSan147 committed:
credit fatalgoth

RaSan147 committed:
credit windjackz

RaSan147 committed:
Update pixijs link README.md

RaSan147 committed:
Update README.md

RaSan147 committed:
add callback function to motion and speak

RaSan147 committed:
fix Cannot find type definition file for 'offscreencanvas'.

RaSan147 committed:
Automated report

RaSan147 committed:
Update README.md

RaSan147 committed:
patch node.env.NODE_ENV issue (action failed to update)

RaSan147 committed:
Fix #6 for crossOrigin docs

RaSan147 committed:
Disable Force priority if a audio is being played

also updated package+version info
fixed NodeEnv issue

RaSan147 committed:
Update version number

RaSan147 committed:
Update cubism

RaSan147 committed:
update dist

RaSan147 committed:
Update README.md

RaSan147 committed:
Revert and Force disable next motion while speaking

xiaoqiang1999 committed:
update readme

xiaoqiang1999 committed:
update readme

RaSan147 committed:
Merge pull request #10 from xiaoqiang1999/master

Update Readme

xiaoqiang1999 committed:
update README

RaSan147 committed:
Merge pull request #11 from xiaoqiang1999/master

Improve example code quality in update README ZH and EN

xiaoqiang1999 committed:
update README.zh

RaSan147 committed:
Merge pull request #13 from xiaoqiang1999/master

update README.zh

RaSan147 committed:
remove unused dep

shiftonetothree committed:
feat(SoundManager.ts): fix audio memory leak

shiftonetothree committed:
fix(SoundManager.ts): lint file

RaSan147 committed:
Updated GH actions versions

shiftonetothree committed:
fix(SoundManager.ts): fix for unit test

shiftonetothree committed:
feat(vite.config.ts SoundManager.test.ts): add --autoplay-policy=no-user-gesture-required

shiftonetothree committed:
feat(MotionManager.*): fix tescase when audio can play

RaSan147 committed:
Merge pull request #23 from shiftonetothree/fix/fix-audio-memory-leak

Fix/fix audio memory leak

shiftonetothree committed:
feat(SoundManager.ts makeTestSound* stability.test.ts): improve stability when model speak over 1000 times

shiftonetothree committed:
feat(make-test-sound): rename file

RaSan147 committed:
removed logging onFinish & lint a bit

RaSan147 committed:
Merge pull request #24 from shiftonetothree/fix/fix-audio-memory-leak

Fix/fix audio memory leak

RaSan147 committed:
Update .gitmodules

to test CI success

RaSan147 committed:
Run with missing lib installed

RaSan147 committed:
bring back

RaSan147 committed:
Automated report

RaSan147 committed:
test updates

RaSan147 committed:
Update test.yml

RaSan147 committed:
Revert patch-dts-generator.js

RaSan147 committed:
Update compatibility.test.ts

shiftonetothree committed:
feat(stability.test.ts): fix ci

shiftonetothree committed:
fix(vite.config.ts): fix ci

shiftonetothree committed:
feat(vite.config.ts): set testTimeout to 10s

RaSan147 committed:
Merge pull request #25 from shiftonetothree/CI-test

Improve test case stability in GitHub action

shiftonetothree committed:
fix(vite.config.ts): add missing bundleTestFiles

shiftonetothree committed:
fix(vite.config.ts): set fileParallelism to false

RaSan147 committed:
Merge pull request #26 from shiftonetothree/CI-test

fix(vite.config.ts): add missing bundleTestFiles

RaSan147 committed:
removed submodule download part

shinshin86 committed:
fix(cubism4): skip updateWebGLContext when clipping manager is absent

shinshin86 committed:
fix error of branch or tag could not be found (CI)

RaSan147 committed:
Automated report

RaSan147 committed:
Merge pull request #27 from shinshin86/fix/cubism4-no-clipping-crash

Fix crash when model has no clipping masks
Also fixes #12

RaSan147 committed:
Automated report

mikeyang01 committed:
Update README.md

correct spell miss

RaSan147 committed:
Merge pull request #28 from mikeyang01/patch-1

Update README.md

RaSan147 committed:
Update to v0.5.0-ls-8

Changelog:
Improved stability
Added translated readme
Removed unwanted dep
Fix GitHub action
Updated submodule policy

RaSan147 committed:
Automated report

RaSan147 committed:
update submodule stuffs and build

RaSan147 committed:
Automated report
